ios/rxswift_sherpa.md

```diff
@@ -1,4 +1,122 @@
 # Rx{Swift,Cocoa} Sherpa
+
+*** **This document is under active development** ***
+
+This document contains best practices when using RxSwift, RxCocoa, and other Rx libraries in iOS projects.
+
+## Rx-enabled View Models
+
+### Setting up your Observables
+
+When possible, a ViewModel's observable sequences should be setup in a declarative fashion within the ViewModel's `init` method or in lazily declared properties. This allows you to specify your exposed observables as non-optional immutable `let` properties.
+
+```
+private let user = Variable<User?>(nil)
+let username: Observable<String>
+
+init(identifier: String) {
+  APIClient.user(withIdentifier: identifier).catchErrorJustReturn(nil).bind(to: user).disposed(by: disposeBag)
+  username = user.asObservable.map({ $0.username })
+}
+```
+
+The majority of a ViewModel's exposed observable sequences will be observed by objects in the UI layer. CocoaTouch requires that UI updates be performed on the main thread. Additionally, since these observations should generally continue until the ViewController leaves the screen, the observed sequences should not complete or error. For these reasons, you should use the correct observable type for the given situation:
+
+* Use a `Driver` when data is set by the ViewModel and is read-only from the View's perspective
+* Use a `Variable` when double-binding is required to allow the observing View to update observable sequence's most recent value based on user input
+* Use of `Observable` should be limited to when the initial stream is a immediately available; although, in these cases, a plain swift property may be sufficient
+
+__Open Question: Should we be advocating against using Observable at all?__
+
+### Observing
+
+You should prefer `.bind(to: ...)` or `.drive(...)` over `.subscribe(...)` as these are more declarative and less prone to user error.
+
+**Do this**
+```
+// Binding
+viewModel.titleObservable.bind(to: titleLabel.rx.text).disposed(by: disposeBag)
+// Driving
+viewModel.titleDriver.drive(titleLabel.rx.text).disposed(by: disposeBag)
+```
+**Not this**
+```
+// Subscribing
+viewModel.title.subscribe(onNext: { [weak self] title in self?.titleLabel.text }).disposed(by: disposeBag)
+```
+
+#### Where to perform the bindings?
+
+A ViewModel's observables should be bound to a ViewController's outlets from `viewWillAppear` or in a dependent method. This allows time for the ViewModel to be injected into the ViewController (if necessary) and for the outlets initial value to be set prior to the View being displayed on screen. Additionally, it allows for the disposal and rebinding when the view controller leaves and returns to the screen.
+
+```
+override func viewWillAppear(_ animated: Bool) {
+  super.viewWillAppear(animated)
+  viewModel.title.bind(to: self.rx.title).disposed(by: disposeBag)
+}
+```
```

Comment:
I think we need to explicitly show the disposal as well, or we will end up with a lot of people missing that step.

Reply:
The "Disposing" section down below is probably good enough tho IMO